feat: add ability to configure multiple opal validation tests

slobsv opened the pull request:
## What does this PR do?

(Note, this PR follows feedback from [a previously closed PR](https://github.com/observeinc/.github/pull/12); feel free to take a look at that for more context.)

Adds a ci job to the `terraform-observe-*` ci tests that does OPAL validation and fails in the case that a new commit has an observe resource with bad OPAL (or otherwise fails to be applied). 

What it does in detail:

1. installs terraform
2. checks out the repo
3. iterating through every `.tests/` subdirectory, creates a boilerplate `provider.tf`, and runs `terraform init`. If it does not find any `.tests/` subdirectories, it will fail with a message pointing to a readme for how to write tests. 
4. iterating through every `.tests/` subdirectory, runs `terraform apply` and `terraform plan`.  (applies with `-var name_format="tf_test_${{ github.sha }}$dir%s"` so that each workflow's observe-resources can be identified for each test, if for some unexpected reason they were not to be properly destroyed)
5. iterating through every `.tests/` subdirectory, runs `terraform destroy` (set to run `always()` so that this job doesn't leave orphaned resources in the observe account)

Reasons for keeping steps `3, 4, 5` separate and not joining them: 
- if destroy is not alone, then a failed apply will result in destroy not running (leaves orphaned resources in observe account)
- if init is not alone, then if plan or apply fail, a test `init` may not be run. If a `destroy` is attempted without an `init` then it fails, and subsequent `destroy` commands may not be run. 
- keeping everything separate like this allows us to add a new step after `4` where we run custom assertions, and we'll probably want that to be in its own step for easier troubleshooting when those fail

## Motivation

automatic testing for valid OPAL in ci

## Limitations / Assumptions 

**First**, This test assumes that there is some observe instance that is available for testing with. It only terraforms with a local backend (so no risk of competing terraform backends with multiple people applying on the same observe test instance), and always destroys even if errors occur in the apply step (so no risk of leaving clutter after failed applies).

**Second**, This test introduces 3 new variables which are currently expected as secrets. These include:

1. secrets.TERRAFORM_MODULES_OBSV_TEST_INST_ID
2. secrets.OBSV_USER_HANDLE
3. secrets.OBSV_USER_PSWD

I think these should basically all be org-wide secrets, since I expect we'll just want to use the same test account for them all. Later on, we may introduce another input for datastreams, which we may end up wanting to be repo-specific, but that's for a later time. 

**Third**, tests are created with subdirectories under `.tests/`, and each of these subdirectories contain the terraform files you would need to run a test scenario. For example, you can have a `.tests/default/` directory that would contain the bare minimum needed in its `main.tf`, `variables.tf`, etc. to apply the module with its default variable inputs. 

## Testing

I've tested this check for both successes and failures on [this test repo branch](https://github.com/observeinc/terraform-observe-estib-test1/tree/slobsv/test-opal-val). 
- [Successful run](https://github.com/observeinc/terraform-observe-estib-test1/runs/6617337271?check_suite_focus=true)
- [Failed run](https://github.com/observeinc/terraform-observe-estib-test1/runs/6617312139?check_suite_focus=true#step:2:658) (with bad OPAL intentionally introduced)

jta commented on .github/actions/terraform-observe/opal-validation/action.yaml:
```diff
@@ -0,0 +1,87 @@
+  observe-domain:
+    description: "observeinc.com or observe-staging.com"
+    required: true
+  observe-user-handle: 
```
Use the official nomenclature in tf. `OBSERVE_USER_EMAIL`, `OBSERVE_USER_PASSWORD`.

jta commented on .github/actions/terraform-observe/opal-validation/action.yaml:
```diff
@@ -0,0 +1,87 @@
+          # uses cat since passwords can contain chars that makes sed sad (&\)
+          cat <<EOF > ci_added_provider.tf
+          provider "observe" {
+          customer      = "${{ inputs.observe-customer-id }}"
```
You don't need this. Just ensure that provider is not set and configure the environment variable?

slobsv replied:
wow, didn't realize we could set these with env vars. That's nice!

jta commented on .github/actions/terraform-observe/opal-validation/action.yaml:
```diff
@@ -0,0 +1,87 @@
+          ran_init=true
+        done
+        
+        # there should always be tests!
```
This is too strict for now.

slobsv replied:
sounds good, we can always force tests later

jta commented on .github/actions/terraform-observe/opal-validation/action.yaml:
```diff
@@ -0,0 +1,87 @@
+      shell: bash
+      run: |
+        ran_init=false
+        for dir in .tests/*/ ; do
```
Use a build matrix driven by the result of `ls .tests`. This should make things cleaner.

slobsv replied:
very nice

jta commented on .github/workflows/terraform-observe_scheduler.yaml:
```diff
@@ -31,6 +31,21 @@ jobs:
         env:
           GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
 
+  opal-validation:
```
`test-apply`.

jta commented on .github/actions/terraform-observe/opal-validation/terraform_tests.sh:
```diff
@@ -0,0 +1,55 @@
+}
+
+do_destroy() {
+	go_dir
```
Use terraform's -chdir option instead, it's more robust.

jta commented on .github/workflows/terraform-observe_scheduler.yaml:
```diff
@@ -31,6 +31,54 @@ jobs:
+        id: matrix
+        shell: bash
+        run: |
+          DIRS=$(python -c "import json; import glob; print(json.dumps([x.replace('/main.tf', '') for x in glob.glob('.tests/**/main.tf', recursive=True)]))")
```
Can you just use `ls` instead? That would avoid installing python. You can assume any directory under ./tests is intended as a test.

slobsv replied:
the tricky part was getting bash to format the results of `ls` to a list-friendly string, like `["test1","test2"]`, but eventually i think i got it.

slobsv commented:
Thanks for the feedback (here and in slack)!

Following the feedback, i've made some changes, among them:

1. Renamed this whole thing from "opal-validation" to `terraform-test-apply` since I'm now considering this a "terraform" type test where what we're testing is "apply". OPAL validation is just one of the goals.
2. Moved the bulk of the work of the test to a bash script. This will ultimately be run from a `Makefile` that will be defined in each `terraform-observe-*` repo (and generated from the template repo). I've provided content of this bash script here for reference under `.github/scripts/terraform_tests.sh`. The benefit of distributing the script to all the repos is that it's easy for end-users to locally run the test script themselves with a `make tests TEST_NAME=foo` -- so that they can troubleshoot test failures without having to push through a million github actions
3. Moved to a matrix build of all the test directories
4. Removed the requirement that there be tests (we can add that later)
5. Removed assumption that there will be a `name_format` tfvar -- as a result, it will ultimately fall on the test writer to randomize their names (we'll just need to document this in a "how to write tests" doc in the template repo)
6. Switched to using the observe provider's environment vars instead of needing to make a `provider.tf` 

Some notes:

- Creating tests will require some work, so we'll have to document it, probably in the template repo
- You can see what a sample PR adding new tests would look like [here](https://github.com/observeinc/terraform-observe-estib-test1/pull/19/files) (just everything under `.github/` and `.tests/` and the `Makefile`, plus with an example test run [here](https://github.com/observeinc/terraform-observe-estib-test1/actions/runs/2416666816)
- Any individual repo can opt out of this job by adding the `skip` input with a value of `{"jobs": ["terraform-tests"]}` if they wish -- but if they haven't written any test directories, nothing would fail from this workflow anyway

Big thanks João for all the input and pointers!

jta commented on .github/workflows/terraform-observe_scheduler.yaml:
```diff
@@ -31,6 +31,42 @@ jobs:
+          OBSERVE_CUSTOMER: ${{ secrets.TERRAFORM_MODULES_OBSV_TEST_INST_ID }}
+          OBSERVE_DOMAIN: "observe-staging.com"
+          OBSERVE_USER_EMAIL: ${{ secrets.TERRAFORM_MODULES_OBSERVE_USER_EMAIL }}
+          OBSERVE_USER_PASSWORD: ${{ secrets.TERRAFORM_MODULES_OBSERVE_USER_PASSWORD }}
```
We already have lots of `TERRAFORM_MODULES_`, maybe group these under `TERRAFORM_MODULES_TEST_`.

jta commented on .github/workflows/terraform-observe_scheduler.yaml:
```diff
@@ -31,6 +31,42 @@ jobs:
+
+  terraform-test-apply:
+    needs: get-test-directories
+    if: "!contains(fromJSON(inputs.skip).jobs, 'terraform-tests')"
```
Isn't this condition implicit in `needs`? There's probably some way to avoid expressing the condition twice.

slobsv replied:
oh clever! 

unfortunately from some testing just now, i've learned that if a build matrix is empty, it doesn't just skip the downstream tests, it causes a failure (which will happen frequently until every repo gets its tests written). So while you're right, i can do without the ` if: "!contains(fromJSON(inputs.skip).jobs, 'terraform-tests')"` here, i had to add a new conditional to make sure the build matrix is not empty. 

(Which, fwiw, that conditional looks like this atm: `if: ${{ needs.get-test-directories.outputs.directories != '[]' && needs.get-test-directories.outputs.directories != '' }}`

checks for `[]` if there are no test directories, and for `''` if the `get-test-directories` job is skipped by the calling workflow. )

slobsv commented on .github/workflows/terraform-observe_scheduler.yaml:
```diff
@@ -31,6 +31,42 @@ jobs:
+        uses: observeinc/.github/.github/actions/terraform-observe/terraform-test-apply@main
+        env: 
+          OBSERVE_CUSTOMER: ${{ secrets.TERRAFORM_MODULES_OBSV_TEST_INST_ID }}
+          OBSERVE_DOMAIN: "observe-staging.com"
```
👍

slobsv commented on .github/workflows/terraform-observe_scheduler.yaml:
```diff
@@ -31,6 +31,42 @@ jobs:
+      - name: Validate OPAL with TF apply/destroy
+        uses: observeinc/.github/.github/actions/terraform-observe/terraform-test-apply@main
+        env: 
+          OBSERVE_CUSTOMER: ${{ secrets.TERRAFORM_MODULES_OBSV_TEST_INST_ID }}
```
👍

jta commented on .github/workflows/terraform-observe_scheduler.yaml:
```diff
@@ -31,6 +31,42 @@ jobs:
+        id: matrix
+        shell: bash
+        run: |
+          DIRS="$(find . -path "*.tests/*" ! -path "*.terraform*" -type d )"
```
I think we'll end up having to move this logic into the scripts/terraform_tests.sh script.

slobsv replied:
interesting -- why so?

is it that you'd like users to be able to run all the tests locally at the same time? if yes, we could update the script to support an "all" argument that does this extra logic, while otherwise accepting individual paths. 

That way we can keep the build matrix (which i'd like to do, since that means all the test run in parallel quickly)

jta replied:
A few reasons:
- tests should be reproducible and portable. Having part of the logic in github actions makes it harder to discern what the expected layout is for example. If we ever decide to use jenkins instead, we have to convert this to some other framework. None of that happens if you restrict things to `make test`
- the logic for for the find may become more convoluted, I'm not sure. Right now we're at the limit of reasonable.
- we can parallelize in the bash script or makefile. It's not as pretty, but totally doable.
- our apiserver may have issues with parallelization. It shouldn't, but it might, so we may have to apply serially anyway.

slobsv replied:
oh interesting, thanks for expanding on that! I'll look into parallelizing in bash/makefile, but i'll keep in mind possible limits of the apiserver on parallelization (maybe i can allow for configurable serial/parallel executions or something)

slobsv replied:
oh cool! then i'll just move forward with this for now :) thanks for clarifying!

slobsv commented:
- updated secret names
- made `observe-staging.com` the default value for domain in the script
- updated `make` command to use `make test path/to/test` and tested both locally and in CI
- updated matrix builder to accept all `.tests/` subdirectories even if they're under nested modules, tested successfully in CI ([example results](https://github.com/observeinc/terraform-observe-estib-test1/actions/runs/2424620910))
- updated conditionals for `terraform-test-apply` workflow to make it so that job will pass successfully even if no test directories exist, or if the `get-test-directories` job is skipped by the caller workflow

jta commented on .github/scripts/terraform_tests.sh:
```diff
@@ -0,0 +1,34 @@
+INFO() { echo "::info::$*"; }
+DEBUG() { test -z "$VERBOSE" || echo "::debug::$*"; }
+
+OBSERVE_DOMAIN="${OBSERVE_DOMAIN:-observe-staging.com}"
```
ditch this, it won't be a sane default for external users. It's better for our usecase to be explicit anyway, and since inherit secrets there's no duplication to worry about.

slobsv replied:
👍
